crates/burn-core/src/nn/sequential.rs

```diff
@@ -0,0 +1,92 @@
+/// Create a sequential neural network, similar to numpy's nn.Sequential.
+///
+/// To use this macro, separate your modules into three categories:
+/// - Unit modules: Modules that don't take any parameters (eg. Relu, Sigmoid)
+/// - Modules: Modules that take parameters, but don't have a backend parameter (eg. Dropout, LeakyRelu)
+/// - Backend modules: Modules that take a backend parameter (eg. Linear)
+///
+/// List these classes of modules as comma-separated within classes, then semicolons between, like so:
+/// ```ignore
+/// gen_sequential! {
+///     // No config
+///     Relu,
+///     Sigmoid;
+///     // Has config
+///     DropoutConfig => Dropout,
+///     LeakyReluConfig => LeakyRelu;
+///     // Requires a backend (<B>)
+///     LinearConfig => Linear
+/// }
+/// ```
```

Comment:
I would use pattern matching to differentiate modules that require a backend and config:

```rust
Sequential!(
    Relu, // Without config
    Dropout(DropoutConfig), // With config
    Linear(LinearConfig; B), // With config + Generics
    Custom(; B), // No config + Generics
    Custom2(config; B, A, C), // With config + many generics
)
```

Reply:
While this would be my ideal method for doing this, it would require me to rewrite it as a proc macro since there isn't a great way to differentiate depending on whether a value is present or not (specifically the whole `($cfg$(; $($generic),+))?`) in declarative macros. Separating into multiple blocks allows me to know that the structs in that block need to have `.init()` called if they have a config or `.init(device)` called if they have a backend-dependent config without needing actual Rust code to differentiate them. A proc macro would fix this but would also require a new crate just for the macro and some more advanced parsing techniques.

Additionally, how would multiple generics work? Do all generics need to be unique? If I define `A` as generic across `Custom2` and `Custom3`, is it the same generic? I assume it would be. We would probably designate `B` as reserved and meaning "needs a device passed to it on initialization".

Should I look into rewriting this as a more complicated but more flexible proc macro or should I keep the simpler but slightly more rigid declarative macro?

Reply:
Hmm, I actually don't think the macro helps much, and I don't think we should implement a proc macro. Maybe the real solution would be to implement a trait `Forward` instead of simply having a method. We could then support tuples as sequential layers. The `Forward` trait would be totaly decoupled from the `Module` trait and only used to simplify composing multiple `forward` methods.